### PR TITLE
Adds AlleleDepthPseudoCounts genotype annotation.

### PR DESCRIPTION
**Pull request by vruano**
Similar to AD, the new annotation (DD) captures the depth of each allele supporting evidence
or reads, however it does so by following a variational Bayes approach looking into the
likelihoods rather than applying a fix threshold. 

This turns out to be more robust in some instances.

To get the new non-standard annotation in HC you need to add -A AllelePseudoDepth

### CODE REVIEW

**Comment by fleharty** [comment]
@vruano 
Tests are failing, could you resolve this?
It appears to be in:
org.broadinstitute.hellbender.tools.walkers.annotator.VariantAnnotatorIntegrationTest > testAgainstMutect2 FAILED
    java.lang.IllegalArgumentException: Dirichlet parameters may not be negative
        at org.broadinstitute.hellbender.utils.Utils.validateArg(Utils.java:798)
        at org.broadinstitute.hellbender.utils.Dirichlet.<init>(Dirichlet.java:26)
        at org.broadinstitute.hellbender.tools.walkers.mutect.SomaticLikelihoodsEngine.getEffectiveCounts(SomaticLikelihoodsEngine.java:56)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
+package org.broadinstitute.hellbender.tools.walkers.annotator;
+
+import htsjdk.variant.variantcontext.*;
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:3]: Should these be explicit imports?  I'm not sure.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    private static DecimalFormat DEPTH_FORMAT = new DecimalFormat("#.##");
+    private static DecimalFormat FRACTION_FORMAT = new DecimalFormat("#.####");
+
+    public static final String DEPTH_KEY = "DD";
+    public static final String FRACTION_KEY = "DF";
+
+    public static final VCFFormatHeaderLine DEPTH_HEADER_LINE =
+            new VCFFormatHeaderLine(DEPTH_KEY, VCFHeaderLineCount.R, VCFHeaderLineType.Float, "Allele depth based on Dirchlet posterior pseudo-counts");
+    public static final VCFFormatHeaderLine FRACTION_HEADER_LINE =
+            new VCFFormatHeaderLine(FRACTION_KEY, VCFHeaderLineCount.R, VCFHeaderLineType.Float, "Allele Fraction based on Dirchilet postrior pseudo-counts");
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:37]: Spelling Dirichlet, 
and posterior

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+public final class AllelePseudoDepth implements GenotypeAnnotation {
+
+    private static DecimalFormat DEPTH_FORMAT = new DecimalFormat("#.##");
+    private static DecimalFormat FRACTION_FORMAT = new DecimalFormat("#.####");
+
+    public static final String DEPTH_KEY = "DD";
+    public static final String FRACTION_KEY = "DF";
+
+    public static final VCFFormatHeaderLine DEPTH_HEADER_LINE =
+            new VCFFormatHeaderLine(DEPTH_KEY, VCFHeaderLineCount.R, VCFHeaderLineType.Float, "Allele depth based on Dirchlet posterior pseudo-counts");
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:35]: Spelling Dirichlet

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+@DocumentedFeature(groupName= HelpConstants.DOC_CAT_ANNOTATORS, groupSummary=HelpConstants.DOC_CAT_ANNOTATORS_SUMMARY, summary="Total depth of coverage per sample and over all samples (DP)")
+public final class AllelePseudoDepth implements GenotypeAnnotation {
+
+    private static DecimalFormat DEPTH_FORMAT = new DecimalFormat("#.##");
+    private static DecimalFormat FRACTION_FORMAT = new DecimalFormat("#.####");
+
+    public static final String DEPTH_KEY = "DD";
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:31]: Should DEPTH_KEY and FRACTION_KEY be in, GATKVCFConstants?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            new VCFFormatHeaderLine(FRACTION_KEY, VCFHeaderLineCount.R, VCFHeaderLineType.Float, "Allele Fraction based on Dirchilet postrior pseudo-counts");
+
+    @Argument(fullName = "dirichlet-prior-pseudo-count",
+              doc = "Pseudo-count used as prior for all alleles. The default is 1.0 resulting in a flat prior")
+    @Advanced
+    public double prior = 1.0;
+
+    @Argument(fullName = "dirichlet-keep-prior-in-count",
+              doc = "By default we don't keep the prior use in the output counts ase it makes it easier to interpret" +
+                      "this quanity as the number of supporting reads specially in low depth sites. We this toggled the prior is included")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:46]: spelling, quantity

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    @Argument(fullName = "pseudo-count-weight-decay-rate",
+              doc = "A what rate the weight of a read decreases base on its informativeness; e.g. 1.0 is " +
+                      "linear decay (default), 2.0 is for quadratic decay",
+              minValue = 0.0)
+    public double weightDecay = 1.0;
+
+    private static final List<String> KEYS = Arrays.asList(DEPTH_KEY, FRACTION_KEY);
+    private static final List<VCFCompoundHeaderLine> HEADER_LINES = Arrays.asList(DEPTH_HEADER_LINE, FRACTION_HEADER_LINE);
+
+    private double[][] priorPseudoCounts = new double[10][]; // possibly won't ever go beyond 10.
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:58]: This seems really dangerous, what's the justification for 10?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        return HEADER_LINES;
+    }
+
+    @Override
+    public List<String> getKeyNames() {
+        return KEYS;
+    }
+
+    @Override
+    public void annotate(final ReferenceContext ref, final VariantContext vc,
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:71]: Could you add some javadoc for this method?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            lkMatrix.walkInOptimizedOrder(log10ToLnTransformer);
+        } else {
+            lkMatrix = sampleMatrixForAlleles.asRealMatrix();
+        }
+        return lkMatrix;
+    }
+
+    private double[] calculateWeights(RealMatrix lkMatrix) {
+        if (weightDecay <= 0) {
+            return null;
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:129]: Would it be more sensible to throw an exception here?

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java [hunk 9]
```diff
@@ -113,4 +116,35 @@
      * @param offset the copy offset within the destination allele
      */
     void copyAlleleLikelihoods(final int alleleIndex, final double[] dest, final int offset);
+
+    /**
+     * Returns this matrix as a {@link RealMatrix}.
+     * <p>
+     *     Changes int he return matrix will affect the content of
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java:123]: "in the"
and of what?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        } else {
+            final double[] result = new double[lkMatrix.getColumnDimension()];
+            for (int i = 0; i < result.length; i++) {
+                double best = lkMatrix.getEntry(0, i);
+                double secondBest = Double.NEGATIVE_INFINITY;
+                for (int j = 1; j < lkMatrix.getRowDimension(); j++) {
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:135]: This makes me really nervous.  Why, does i start with 0, and j with 1?

**Review by fleharty — CHANGES_REQUESTED** [review]
@vruano I have a few minor nits.  I think we should aim to get this in tomorrow so that it is part of the release.  Rob and Richard have asked that we get this in this release if possible.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 10 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:135]: i is standard 
j if you look at the other two previous lines in the outer for loop (```double best = ...``` and ```double secondBest = ```) the work for j == 0 is done, that is why we jump straight to j = 1

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 6 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:58]: we resize this array as needed, 10 is magic number that would cover 99% of cases (10 alleles or less).
I can give it a name.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 8 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:129]: if less than 0.0, we can do that.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 7 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:71]: I would say that the doc is inherited from the parent class. Is that sufficient.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 10 above)
Comment (fleharty) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:138]: I see, ty.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java (hunk 7 above)
Comment (fleharty) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AllelePseudoDepth.java:72]: Yep, the @inheritDoc is good.

**Review by fleharty — CHANGES_REQUESTED** [review]
This looks good, though, I'm uncomfortable with, 
INITIAL_PRIOR_CACHE_MAX_ALLELE = 10.
The reason is that in high depth somatic samples I we sometimes see more than 10 alleles show up due to homopolymer runs.  At high depth sometimes you'll see an insertion of 1 A, AA, AAA, etc on out to like 14 As.

This is rare, but it does happen.

Are there any solutions where we can easily deal with a general number of alleles?  Sorry, I know it's a pain.

**Comment by vruano** [comment]
Test pass, @fleharty want to merge?

**Review by fleharty — APPROVED** [review]
Thanks Valentin!

**Pull request merged**